pkg/networkservice/common/connect/server.go

```diff
@@ -21,247 +19,112 @@ package connect
+	conn, err := next.Server(ctx).Request(ctx, request)
 	if err != nil {
-		// We not succeed, just remove connection by id, if it was existing one.
-		<-c.executor.AsyncExec(func() {
-			if request.Connection.Id != "" {
-				// Delete from CE
-				delete(ce.connections, request.Connection.Id)
-				// Delete from global list
-				delete(c.connections, request.Connection.Id)
-			}
-		})
 		return nil, err
 	}
+	// Copy Context and Path from client to response from server
+	conn.Context = clientConn.Context
+	conn.Path = clientConn.Path
 
-	// Copy conn to request and pass to next one
-	if conn.GetContext() != nil {
-		request.Connection.Context = conn.GetContext()
-	}
-	// we succeed, update connection and remove pending state
-	<-c.executor.AsyncExec(func() {
-		ce.connections[conn.Id] = conn
-
-		// Also update global connection map
-		c.connections[conn.Id] = clientURL
-	})
-
-	return next.Server(ctx).Request(ctx, request)
+	// Return result
+	return conn, err
 }
 
 func (c *connectServer) Close(ctx context.Context, conn *networkservice.Connection) (*empty.Empty, error) {
-	clientURL, err := c.clientURL(ctx, conn)
-	if err != nil {
-		return nil, err
+	var clientErr error
+	client, _ := c.clientsByID.Load(conn.GetId())
+	if client == nil {
+		return &empty.Empty{}, nil
 	}
-	// Find or Create clientEntry, with add pending.
-	var ce *clientEntry
-	ce, err = c.findOrCreateClient(ctx, clientURL, conn)
-	if err != nil {
-		// In case of error, no pending state is propogated.
-		return nil, err
-	}
-
-	// Call client close
-	_, clientErr := ce.client.Close(ctx, conn)
-	// remove connection from list
-	<-c.executor.AsyncExec(func() {
-		delete(ce.connections, conn.Id)
-		delete(c.connections, conn.Id)
-		// Close client if there is no more users for it.
-		c.closeClient(ctx, ce)
-	})
+	_, clientErr = client.Close(ctx, conn)
 	rv, err := next.Server(ctx).Close(ctx, conn)
 	if clientErr != nil && err != nil {
 		return rv, errors.Wrapf(err, "errors during client close: %v", clientErr)
 	}
 	if clientErr != nil {
 		return rv, errors.Wrap(clientErr, "errors during client close")
 	}
-
 	return rv, err
 }
 
-func (c *connectServer) findOrCreateClient(ctx context.Context, clientURI *url.URL, conn *networkservice.Connection) (ce *clientEntry, err error) {
-	connectionExists := false
-	<-c.executor.AsyncExec(func() {
-		ce, connectionExists = c.clients[clientURI.String()]
-		if !connectionExists {
-			ce = &clientEntry{
-				clientURI:   clientURI,
-				ready:       make(chan struct{}),
-				connections: map[string]*networkservice.Connection{},
-			}
-			// Put/Update connection
-			c.clients[clientURI.String()] = ce
+func (c *connectServer) client(ctx context.Context, conn *networkservice.Connection) networkservice.NetworkServiceClient {
+	// Fast path, if we have a client for this conn.GetId(), use it
+	client, _ := c.clientsByID.Load(conn.GetId())
+
+	// If we didn't find a client, we fall back to clientURL
+	if client == nil {
```

Comment:
Imagine the case when client called two times Reuqest in parallel in this case this condition for one client can be true twice. When expected that the first call added to 'cache' map client and the second call should just load this value. I also think that this problem can be detected by a test with few requests in separate goroutines.

Reply:
We actually have tests with Requests in different go routines :) (10 of them).

You are correct that we could, in theory, have client == nil be true here for multiple nearly simultaneous Requests with the same Conn.Id(). See comment below for why we don't care :)

Reply:
Also updated the comments in the file to reflect all of this to make sure its clear to subsequent readers :)

Reply:
Looks good to me with the flag `loaded` it will work correctly :)